parquet-column/src/main/java/org/apache/parquet/io/MessageColumnIO.java

```diff
@@ -294,17 +301,33 @@ private void writeNullForMissingFieldsAtCurrentLevel() {
 
     private void writeNull(ColumnIO undefinedField, int r, int d) {
       if (undefinedField.getType().isPrimitive()) {
-        columnWriter[((PrimitiveColumnIO)undefinedField).getId()].writeNull(r, d);
+        columnWriter[((PrimitiveColumnIO) undefinedField).getId()].writeNull(r, d);
       } else {
-        GroupColumnIO groupColumnIO = (GroupColumnIO)undefinedField;
-        writeNullToLeaves(groupColumnIO, r, d);
+        GroupColumnIO groupColumnIO = (GroupColumnIO) undefinedField;
+        // only cache the repetition level, the definition level should always be the definition level of the parent node
+        cacheNullForGroup(groupColumnIO, r);
       }
     }
 
-    private void writeNullToLeaves(GroupColumnIO group, int r, int d) {
-      for(ColumnWriter leafWriter: groupToLeafWriter.get(group)) {
-        leafWriter.writeNull(r,d);
+    private void cacheNullForGroup(GroupColumnIO group, int r) {
+      if (!groupNullCache.containsKey(group)) {
+        groupNullCache.put(group, new IntArrayList());
       }
+      groupNullCache.get(group).add(r);
+    }
+
+    private void writeNullToLeaves(GroupColumnIO group) {
+      IntArrayList nullCache = groupNullCache.get(group);
+      if (nullCache == null || nullCache.isEmpty())
+        return;
+
+      int parentDefinitionLevel = group.getParent().getDefinitionLevel();
+      for (ColumnWriter leafWriter : groupToLeafWriter.get(group)) {
+        for (int repetitionLevel : groupNullCache.get(group)) {
```

Comment:
here you call `groupNullCache.get(group)` once per leafWriter but group is constant, and you've already stored it in nullCache, so this can be:

```
int parentDefinitionLevel = group.getParent().getDefinitionLevel();
for (ColumnWriter leafWriter : groupToLeafWriter.get(group)) {
  for (int repetitionLevel : nullCache) {
    leafWriter.writeNull(repetitionLevel, parentDefinitionLevel);
  }
}
```